### PR TITLE
New utilities

### PR DESCRIPTION
**Pull request by dalcinl**

### CODE REVIEW

**Comment by dalcinl** [comment]
@jsquyres Looks like Open MPI 4.1.0 introduced changes in the handling of alignment of struct datatypes. This MR has failures related to those changes. These failures only show up on macOS because Linux builds are using previous releases. I tested v4.1.0 on my Linux workstation, and I can reproduce the failures. Everything runs fine with v4.0.4. This is either a regression in Open MPI, or bad assumptions in my test code that now fails after Open MPI fixed previous bugs. FYI, all these tests pass with MPICH.

Do you remember any important changes/fixes to the struct datatype machinery in between the last two minor releases v4.0 and v4.1? What's your gut feeling? A regression in Open MPI or wrong mpi4py test code?

**Comment by dalcinl** [comment]
Open MPI issue reported upstream in open-mpi/ompi#8560. I'm going to skip the failing tests to get this MR moving forward.

**Comment by dalcinl** [comment]
@leofang I know you are really busy, but perhaps you want to give a quick look at the new documentation this PR adds.

**Review by leofang — COMMENTED** [review]
Thanks @dalcinl. I've been reading this intermittently since this morning. Let me leave my first batch of questions and I'll continue to finish `mpi4py.util.pkl5` later tonight, sorry!

File: docs/source/usrman/mpi4py.util.dtlib.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.dtlib.rst:12]: Well, the thing is coded to accept a `np.dtype` instance, and not things like `np.float32` or type strings like `"i"`. But this should be trivial to support. The the arg type would be `DTypeLike`.

Regarding using more descriptive names, why stop there? We can name it `numpy_dtype_to_mpi_datatype()` :sweat_smile:. The module is named `dtlib`, it is explicitly documented to do conversion between NumPy and MPI datatypes. 

But perhaps you have a point. Users love to criticize for lack of documentation, but at the same time, if documentation is provided, they do not care to read it. Let me think a bit more about it, but looks like I'll follow your advice.

File: docs/source/usrman/mpi4py.util.dtlib.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.dtlib.rst:16]: Well, regarding your first question above, in this case the return type is `numpy.dtype`.

File: src/mpi4py/util/dtlib.pyi
Comment (dalcinl) [reply src/mpi4py/util/dtlib.pyi:2]: Good catch! I'll fix it by amending the commit.

File: src/mpi4py/util/dtlib.py
Comment (dalcinl) [reply src/mpi4py/util/dtlib.py:9]: Yes, but that's also the case with  `mpi4py.futures`, the new `mpi4py.util.pkl5`, and many other third-party packages that uses `mpi4py.MPI`. Remember we have the `mpi4py.rc` mechanism to avoid initialization. BTW, I recently added support for controlling `mpi4py.rc` via environment variables in e12723d (of course, I forgot to document it).

File: src/mpi4py/util/dtlib.py
Comment (dalcinl) [reply src/mpi4py/util/dtlib.py:130]: Yes, you get what would be needed to construct a `numpy.dtype`. You may argue that this has little use, but this rather small hunk of code allows for some testing even if numpy is not installed.

File: setup.py
Comment (dalcinl) [reply setup.py:484]: No, this will not turn numpy in a hard test dependency. The tests are carefully coded to run without NumPy. Of course, a few tests are skipped, and others are not as meaningful and exhaustive given that the `numpy.dtype` -> `MPI.Datatype` conversion is not available.

File: docs/source/usrman/mpi4py.util.dtlib.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.dtlib.rst]: I've implemented your renaming suggestions. Now we can pass anything dtype-like to `from_numpy_dtype()`:

```python
>>> from mpi4py.util.dtlib import from_numpy_dtype
>>> duptype = from_numpy_dtype('d')
>>> info = duptype.decode()
>>> info
(<mpi4py.MPI.Datatype object at 0x7fa8b91442f0>, 'DUP', {})
>>> info[0].Get_name()
'MPI_DOUBLE'
```

Comment (leofang) [reply docs/source/usrman/mpi4py.util.dtlib.rst]: This looks great, thanks Lisandro! 
> Regarding using more descriptive names, why stop there? We can name it `numpy_dtype_to_mpi_datatype()` 😅. The module is named `dtlib`, it is explicitly documented to do conversion between NumPy and MPI datatypes.

I know...yeah it's always a tough design decision. My suggestion was based on the fact that many GPU libraries have similar API names (`to_numpy`, `asnumpy`, etc) which focus on device-host transfer. I know we don't do/need it here, but it'd be a bit less intuitive on its purpose these days...So thanks for making it a bit clearer.

File: docs/source/usrman/mpi4py.util.dtlib.rst
Comment (leofang) [reply docs/source/usrman/mpi4py.util.dtlib.rst]: Oh ok, then there's no ambiguity in this case, good to know.

File: src/mpi4py/util/dtlib.py
Comment (leofang) [reply src/mpi4py/util/dtlib.py:9]: I didn't know we accept env vars either! I can work on the doc later, I think we'll need a standalone page for env vars `MPI4PY_RC_*`.

File: setup.py
Comment (leofang) [reply setup.py:484]: OK you are right about testing. But how about generating stubs? I noticed `src/mpi4py/util/dtlib.pyi` imports numpy, so looks like for the purpose of generating docs numpy is a dependency? (I'm totally fine with it, just curious.)

Comment (dalcinl) [reply setup.py:484]: AFAICT, documentation generation does not depend on *.pyi stub files, although perhaps recent Sphinx versions (or third-party extensions) look at them, I really do not know. Stub files are about type checking and not about doc generation. Those using type-checkers will need NumPy installed, of configure things to ignore a missing numpy. Remember that type-checking is "static", I cannot protect `import numpy` with try/except. 

FYI, the only *.pyi stubs file that is autogenerated in full is `mpi4py/MPI.pyi` using `conf/mpistubgen.py`. All the other stub fiels where either written by hand or autogenerated (with mypy's stubgen script) and heavily tweaked by hand afterwards.

File: src/mpi4py/util/pkl5.py
Comment (leofang) [inline src/mpi4py/util/pkl5.py:380]: Can we define `Request.__del__` to call `Free` here so that it gets cleaned up when out of scope? At least you did this here (`del_Request` calls `MPI_Request_Free`, like `Free` does): https://github.com/mpi4py/mpi4py/blob/bb5289d1531329b5ea28780996b51399ddf72200/src/mpi4py/MPI/Request.pyx#L13-L15

**Review by leofang — COMMENTED** [review]
I have a couple of questions on `mpi4py.util.pkl5` 😅

File: setup.py
Comment (leofang) [reply setup.py:484]: > FYI, the only *.pyi stubs file that is autogenerated in full is `mpi4py/MPI.pyi` using `conf/mpistubgen.py`. All the other stub fiels where either written by hand or autogenerated (with mypy's stubgen script) and heavily tweaked by hand afterwards.

Yeah I just realized it...Sorry for missing that until now. I feel so sad that you have to maintain all other `*.pyi` files by hand...Is it a necessary evil?

Comment (dalcinl) [reply setup.py:484]: Well, I believe NumPy folks are following the exact same approach. The hard part is to get the stubs the first time, afterwards maintenance is relatively easy. I still have to setup mypy's stubtest linter (and other linting steps like running pylint) on CI. I'll do it shortly after we merge this PR.

PS: Back in the days, when I was a junior, Python folks bragged about duck typing and that Python was awesome because you did not have to declare variables and its types. But the usual inability of humans of writing a couple lines of code without introducing a bug gave a dosis of reality and put the community in check.  Now the key players are advocating typing and type checkers and all that. However, there is a problem... aside from the compatibility issues (Python 2), the yet-evolving API, and the extra dependencies you may need to inject in your project (e.g. `typing_extensions`), the cruel fact is that Python typing is still an afterthought. If you do not want to maintain *.pyi files, then you have to add typing in your regular Python code. And then you open the door to tons of nasty typing-related code with exact zero-benefit at runtime, all you get is a sort of linter on steroids. Please go to NumPy sources and look at the *.pyi files. If you wanted to merge/combine all that mess on the original *.py sources, you can see how much noise you add to otherwise perfectly good looking Python code.

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.pkl5.rst:37]: I said *... previous Python **3** releases ...* and not *... previous Python releases ...*. 

So an HPC group (HPC!!! what a surprise!!!) don't want to pay the price of moving on while keep spending our time supporting them? We give them working code. I'm not willing to spend my time documenting/clarifying Python 2 support, not after Python 2 is officially dead.

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:37]: If pickle 5 is not available, you still can take advantage of the large-array support. However, this is Python 3 only. IIRC, Python 2 cannot generate pickle streams larger than 2GiB. Anyway, overall, all of this module runs fine on Python 2 (all tests pass), any limitations come from the Python runtime and not mpi4py. I don't see any strong reason to raise errors.

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.pkl5.rst:131]: Maybe. 

Perhaps we could say `-> NoReturn` [link](https://docs.python.org/3/library/typing.html#typing.NoReturn). In fact, the method exists, but raises. However, I agree this is documentation and not typing, so you have a point. 

Or we should actually say something explicit like "this method cannot be supported and raises".

What do you think?

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:503]: I could not figure out a way. The problem is that to issue an `MPI_Irecv()` call you need an upper bound of the message size. 

The normal `irecv()` method uses (by default) a 32K buffer (our you pass a buffer size, or your own scratch buffer space) as a guess, and if your guess is wrong, MPI will complain about a truncated message. Perhaps I should try do something similar here. I think when I started I figured out there was an issue with that approach in here, but maybe I was wrong. Let me think a bit more about it.

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:380]: No, sorry, all Cython  `__dealloc__` methods in mpi4py do nothing, the `flags` slot is always zero. The implementation is half-way, that's the reason `del_Request()` is implemented but effectively unused. For now, until I figure out a robust way to improve things, users are responsible of deallocating every new MPI handle then create. Very unpythonic, but I'm not a fan of promising something I cannot deliver. Most MPI implementations use refcounting to manage MPI handles, unfortunately, that machinery is not exposed (and certainly not part of the MPI standard).

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:37]: Oh, you are not saying that we should raise and error, but rather asking whether that would happen. Is that correct? If yes, then the answer is no, no error is raised, this module works and passes all tests with pickle < 5 and even under Python 2 (although in this case it this module would likely be of little practical benefit for users).

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:67]: I don't understand your question. If pickle 5 is not available, `THRESHOLD` is irrelevant (not used), and everything is pickled in-band. In other words, if pickle 5 is not available, then the `dumps()` method returns `data, bufs` with all of the pickled data in `data` an an empty `bufs` list.

Comment (leofang) [reply src/mpi4py/util/pkl5.py:67]: Ah, I should have been more explicit, sorry. I meant why did we pick 0.25 MiB here and nothing some larger (or smaller) values?

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (leofang) [reply docs/source/usrman/mpi4py.util.pkl5.rst:37]: Well, if I understand you correctly, then unfortunately we still have a one-time cost to pay. AFAIK we never declared mpi4py is Python 3 only -- at least it's not in the 3.0.2 announcement in the mailing list (and I don't find one for 3.0.3), and last time I checked we were still inquiring feedbacks about Python 2.7: https://groups.google.com/g/mpi4py/c/-FufUsS3jOU/m/BXm-nPxDEgAJ. I think for the purpose here it wouldn't hurt to add one sentence along this line "this feature is not supported on Python 2".

File: src/mpi4py/util/pkl5.py
Comment (leofang) [reply src/mpi4py/util/pkl5.py:37]: > the answer is no, no error is raised, this module works and passes all tests with pickle < 5 and even under Python 2 (although in this case it this module would likely be of little practical benefit for users).

Yes, that's what I'm wondering, thanks 🙂

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (leofang) [reply docs/source/usrman/mpi4py.util.pkl5.rst:131]: I don't have strong opinion here, but perhaps being explicit might save users a little time.

File: src/mpi4py/util/pkl5.py
Comment (leofang) [reply src/mpi4py/util/pkl5.py:380]: OK, thanks for clarifying. 
> users are responsible of deallocating every new MPI handle then create

Perhaps we better mention this somewhere in the docs. Let me create an issue to track this.

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:67]: Oh, sorry, I totally misunderstood your question. No, this is not a limitation of pickle 5. If you use out-of-band buffers, you send/receive more messages. If messages are small, latency dominates. So, in my admittedly very limited testing on my workstation, the 0.25 MiB threshold was the value where out-of-band communication started to pay off in terms of wall-clock time. My test used a shared memory channel, which is very low latency. Perhaps we should do more exhaustive testing in the future. Anyway, this should not delay the merge of this MR, if needed, we just need to update this single line with a better default value.

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.pkl5.rst:37]: Yes, you asked for feedback, a not a single soul cared to answer. What should we take of that? Python 2 is effectively dead :wink:?  OK, I'll add a note saying that while the code runs fine on Python 2, the lack of pickle 5 and other limitations of such an old runtime gives no actual benefit (no out-of-band communication, no large-array support).

Comment (leofang) [reply docs/source/usrman/mpi4py.util.pkl5.rst:37]: > Yes, you asked for feedback, a not a single soul cared to answer. What should we take of that? Python 2 is effectively dead 😉?

Yeah it is both sad and good that no one replies...Sad because of the little "community" engagement, and good because we can do whatever we'd like to 🎉

File: src/mpi4py/util/pkl5.py
Comment (leofang) [reply src/mpi4py/util/pkl5.py:67]: Thanks for clarifying. I am just curious about this pick and have no intention to block this PR because of this single line 🙂 I think if there's a need we can revisit later, possible allowing it to be controlled by an env var?

File: src/mpi4py/util/pkl5.py
Comment (leofang) [reply src/mpi4py/util/pkl5.py:503]: The most silly workaround I can think of is to do a blocking send/recv to send the buffer size, in the hope that the overhead is amortized by subsequent large, nonblocking transfer, but I am sure you will think of a better way 😅

But, again, the feature is complete and we can always revisit later. This question should be nonblocking too. We just need to [document](https://github.com/mpi4py/mpi4py/pull/29#discussion_r593377553) this.

**Review by leofang — APPROVED** [review]
Thanks for spending time on working on this and answering my questions, Lisandro! This looks very nice. There are one or two loose ends, mostly on documentation I think. Feel free to merge anytime you like.

File: src/mpi4py/util/pkl5.py
Comment (dalcinl) [reply src/mpi4py/util/pkl5.py:503]: No, no, these tricks do not work reliably! A nonblocking operation has to be truly nonblocking, otherwise you may deadlock. For example, look at the following two pieces of code:

```python
# version 1
comm.ssend(sobj, dest=comm.rank)
req = comm.irecv(source=comm.rank)
robj = req.wait()
```

```python
# version 2
req = comm.irecv(source=comm.rank)
comm.ssend(sobj, dest=comm.rank)
robj = req.wait()
```

A proper implementation of `irecv()` should be such that both versions work as expected. But if you add an initial blocking step, then both will deadlock. In short, better safe than sorry, do not promise what you cannot deliver.

PS: Now I remembered all the issues. `irecv()` cannot possibly be implemented without using polling on a separate thread, and even if I were willing do it, then I cannot make things thread-safe, other users threads would compete to receive messages, completely screwing things up, and there is no way to use a lock without bloking other `recv()` operations (and then,  what the point of using nonblocking receive would be?).

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.pkl5.rst:37]: I'm going to extend the `pickle5` note with the following sentences. Are you OK with it?

```
   The :mod:`pickle5` backport is not available for Python 2. Using the
   :mod:`mpi4py.util.pkl5` module on a Python 2 runtime provides no benefits
   and may hurt communication performance.
```

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (dalcinl) [reply docs/source/usrman/mpi4py.util.pkl5.rst:131]: I'm going to add the following admonition:

```
      .. warning:: This method cannot be supported reliably
                   and raises :exc:`RuntimeError`.
```

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (leofang) [reply docs/source/usrman/mpi4py.util.pkl5.rst:38]: LGTM!

File: docs/source/usrman/mpi4py.util.pkl5.rst
Comment (leofang) [reply docs/source/usrman/mpi4py.util.pkl5.rst:135]: Sounds good 👍

File: src/mpi4py/util/pkl5.py
Comment (leofang) [reply src/mpi4py/util/pkl5.py:503]: You are absolutely right. Having blocking send/recv would break the semantics and cause deadlocks. For some reason I kept thinking this blocking pair can be hidden safely somewhere...

**Pull request merged**
